conductor-client/src/main/kotlin/com/openlattice/data/jobs/DataDeletionJob.kt

```diff
@@ -0,0 +1,524 @@
+
+    companion object {
+        private const val BATCH_SIZE = 10_000
+        private val ALL_PARTITIONS = (0..257)
+    }
+
+    @Transient
+    private lateinit var hds: HikariDataSource
+
+    @Transient
+    private lateinit var byteBlobDataManager: ByteBlobDataManager
+
+    @Transient
+    private lateinit var entitySets: IMap<UUID, EntitySet>
+
+    @Transient
+    private lateinit var entityTypes: IMap<UUID, EntityType>
+
+    @Transient
+    private lateinit var propertyTypes: IMap<UUID, PropertyType>
+
+    override fun initialize() {
+        state.totalToDelete = getTotalToDelete()
+    }
+
+    override fun processNextBatch() {
+        val entityDataKeys = getBatchOfEntityDataKeys()
+
+        if (entityDataKeys.isEmpty()) {
+            hasWorkRemaining = false
+            publishJobState()
+            return
+        }
+
+        var edgeBatch = getBatchOfEdgesForIds(entityDataKeys)
+        while (edgeBatch.isNotEmpty()) {
+            deleteEntities(edgeBatch)
+            deleteEdges(edgeBatch)
+            edgeBatch = getBatchOfEdgesForIds(entityDataKeys)
+        }
+
+        state.numDeletes += deleteEntities(entityDataKeys)
+        cleanUpBatch(entityDataKeys)
+        publishJobState()
+    }
+
+    override fun updateProgress() {
+        if (state.totalToDelete > 0) {
+            progress = ((100 * state.numDeletes) / state.totalToDelete).toByte()
+        }
+    }
+
+    @JsonIgnore
+    private fun getTotalToDelete(): Long {
+        state.entityKeyIds?.let {
+            return it.size.toLong()
+        }
+
+        return hds.connection.use { connection ->
+            connection.prepareStatement(GET_ENTITY_SET_COUNT_SQL).use { ps ->
+                ps.setObject(1, state.entitySetId)
+                ps.executeQuery().use { rs ->
+                    if (rs.next()) {
+                        rs.getLong(1)
+                    } else {
+                        0
+                    }
+                }
+            }
+        }
+    }
+
+    @JsonIgnore
+    private fun getBatchOfEntityDataKeys(): Set<EntityDataKey> {
+        if (state.entityKeyIds != null) {
+            return state.entityKeyIds!!.take(BATCH_SIZE).mapTo(mutableSetOf()) { EntityDataKey(state.entitySetId, it) }
+        }
+        return BasePostgresIterable(PreparedStatementHolderSupplier(hds, getIdsBatchSql()) {
+            it.setObject(1, state.entitySetId)
+            it.setArray(2, PostgresArrays.createIntArray(it.connection, state.partitions))
+        }) {
+            ResultSetAdapters.entityDataKey(it)
+        }.toSet()
+    }
+
+    private fun cleanUpBatch(entityDataKeys: Set<EntityDataKey>) {
+        val entityKeyIds = entityDataKeys.mapTo(mutableSetOf()) { it.entityKeyId }
+        PostgresLinkingQueryService.deleteNeighborhoods(hds, state.entitySetId, entityKeyIds)
+
+        if (state.entityKeyIds == null) {
+            return
+        }
+
+        state.entityKeyIds!!.removeAll(entityKeyIds)
```

Comment:
I don't think you need the `!!` here 🤔

Reply:
and on 143 as well

Reply:
They are needed (I think since the state.entityKeyIds value could, in theory, change in between these calls) but I can make this more kotliny